File: src/main/java/at/hrechny/predictionsbot/exception/interceptor/ExceptionInterceptor.java
package at.hrechny.predictionsbot.exception.interceptor;

import at.hrechny.predictionsbot.service.telegram.TelegramService;
import io.micronaut.aop.MethodInterceptor;
import io.micronaut.aop.MethodInvocationContext;
import jakarta.inject.Singleton;

@Singleton
public class ExceptionInterceptor implements MethodInterceptor<Object, Object> {

  private final TelegramService telegramService;

  public ExceptionInterceptor(TelegramService telegramService) {
    this.telegramService = telegramService;
  }

  @Override
  public Object intercept(MethodInvocationContext<Object, Object> context) {
    try {
      return context.proceed();
    } catch (RuntimeException exception) {
      telegramService.sendErrorReport(exception);
      throw exception;
    }
  }
}

File: src/main/kotlin/at/hrechny/predictionsbot/exception/interceptor/ExceptionInterceptor.kt
package at.hrechny.predictionsbot.exception.interceptor

import at.hrechny.predictionsbot.service.telegram.TelegramService
import io.micronaut.aop.MethodInterceptor
import io.micronaut.aop.MethodInvocationContext
import io.micronaut.context.BeanProvider
import jakarta.inject.Singleton
import org.slf4j.LoggerFactory

@Singleton
class ExceptionInterceptor(
    private val telegramServiceProvider: BeanProvider<TelegramService>,
) : MethodInterceptor<Any, Any> {
    override fun intercept(context: MethodInvocationContext<Any, Any>): Any? {
        try {
            return context.proceed()
        } catch (exception: RuntimeException) {
            try {
                telegramServiceProvider.get().sendErrorReport(exception)
            } catch (reportException: RuntimeException) {
                log.error("Failed to send error report", reportException)
            }
            throw exception
        }
    }

    private companion object {
        val log = LoggerFactory.getLogger(ExceptionInterceptor::class.java)
    }
}
